### PR TITLE
Add document _count API support to Rest High Level Client.

### PR DESCRIPTION
**Pull request by mrdjen**
Add `count()` api method, `CountRequest` and `CountResponse` classes to HLRC. Code in server module is unchanged.

Relates to #27205

### CODE REVIEW

**Comment by hub-cap** [comment]
can you move the `count` package to `core`? Ive told a few people to drop the core (things like search, count, etc..) into the `core` package. None of those PRs have landed yet, so just create it in your PR. im reviewing it now, but i just wanted to ping you with this info so i dont forget :P

**Review by hub-cap — CHANGES_REQUESTED** [review]
Great work so far. We caught you in the middle of some major changes, like the documentation change, so lets clean those up and then ill fire off the CI service against it. <3

**Comment by mrdjen** [comment]
Thanks for comments @hub-cap - I'm addressing these, can you please clarify 

> can you move the count package to core

Do you reffer to `org.elasticsearch.client` package in rest-high-level or shoud I create `core` and move the count package there? (As I don't see `core`)

**Comment by mrdjen** [comment]
OK, I get it, I will create `core` package

**Comment by mrdjen** [comment]
Hi @hub-cap, I've address changes you've requested. Let me know what you think.

Yeah, I see there are some changes going on, this way at least it keeps me in the loop :D, and the XContentTester is usefull.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 1]
```diff
@@ -749,6 +751,29 @@ public final void indexAsync(IndexRequest indexRequest, RequestOptions options,
                 emptySet());
     }
 
+    /**
+     * Executes a count request using the Count API
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: Addressed, let me know what you think.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/count/CountResponse.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                } else {
+                    parser.skipChildren();
+                }
+            }
+        }
+        return new CountResponse(count, terminatedEarly, shardStats);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/count/CountResponse.java]: Addressed, let me know what you think. XContentTester is quite useful!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/count/CountResponse.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                    }
+                } else {
+                    parser.skipChildren();
+                }
+            }
+            return new ShardStats(successfulShards, totalShards, skippedShards, failures.toArray(new ShardSearchFailure[failures.size()]));
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/count/CountResponse.java]: Addressed

File: client/rest-high-level/src/test/java/org/elasticsearch/client/CountIT.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import org.junit.Before;
+
+import java.io.IOException;
+
+import static org.hamcrest.Matchers.greaterThan;
+
+public class CountIT extends ESRestHighLevelClientTestCase {
+
+    @Before
+    public void indexDocuments() throws IOException {
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/CountIT.java]: Moved to SearchIT

File: client/rest-high-level/src/test/java/org/elasticsearch/client/CountRequestTests.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.test.EqualsHashCodeTestUtils.checkEqualsAndHashCode;
+
+//quite similar to SearchRequestTests as CountRequest wraps SearchRequest
+public class CountRequestTests extends ESTestCase {
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/CountRequestTests.java]: Moved

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 6]
```diff
@@ -829,20 +830,20 @@ public void testBulkWithDifferentContentTypes() throws IOException {
...
             bulkRequest.add(new IndexRequest("index", "type", "1").source(singletonMap("field", "value"), XContentType.JSON));
             bulkRequest.add(new UpdateRequest("index", "type", "2")
-                    .doc(new IndexRequest().source(singletonMap("field", "value"), XContentType.JSON))
-                    .upsert(new IndexRequest().source(singletonMap("field", "value"), XContentType.SMILE)));
+                .doc(new IndexRequest().source(singletonMap("field", "value"), XContentType.JSON))
+                .upsert(new IndexRequest().source(singletonMap("field", "value"), XContentType.SMILE)));
             IllegalArgumentException exception = expectThrows(IllegalArgumentException.class, () -> RequestConverters.bulk(bulkRequest));
             assertEquals(
-                    "Mismatching content-type found for request with content-type [SMILE], " + "previous requests have content-type [JSON]",
-                    exception.getMessage());
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: Uhm yes, this should be reverted now

File: docs/java-rest/high-level/search/count.asciidoc [hunk 7]
```diff
@@ -0,0 +1,145 @@
+[[java-rest-high-count]]
+=== Count API
+
+[[java-rest-high-document-count-request]]
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply docs/java-rest/high-level/search/count.asciidoc]: I see. Changed the docs

**Comment by hub-cap** [comment]
Yes there is much in flight now. im having a look at your PR now :)

File: client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java [hunk 8]
```diff
@@ -1233,4 +1235,71 @@ private static void assertSearchHeader(SearchResponse searchResponse) {
         assertEquals(0, searchResponse.getShardFailures().length);
         assertEquals(SearchResponse.Clusters.EMPTY, searchResponse.getClusters());
     }
+
+    //Count API IT tests start
```
Comment (hub-cap) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java]: you dont need these, its a big ole file but we dont do it with other sections

**Review by hub-cap — COMMENTED** [review]
This is looking good. Lets clean up the little nits I mentioned and see if the magic incantations of the tests work.

**Comment by hub-cap** [comment]
Once the nits are addressed please also merge in the latest from master one more time like you've done before. It will be the best way to ensure the tests pass :)

Ill take care of backporting it once its merged. Thanks again for all the work you put in to this!

File: client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java (hunk 8 above)
Comment (mrdjen) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/SearchIT.java]: ok, removed

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 9]
```diff
@@ -485,9 +485,9 @@ public void testRethrottle() {
         List<Tuple<String, Supplier<Request>>> variants = new ArrayList<>();
         variants.add(new Tuple<String, Supplier<Request>>("_reindex", () -> RequestConverters.rethrottleReindex(rethrottleRequest)));
         variants.add(new Tuple<String, Supplier<Request>>("_update_by_query",
-            () -> RequestConverters.rethrottleUpdateByQuery(rethrottleRequest)));
+                () -> RequestConverters.rethrottleUpdateByQuery(rethrottleRequest)));
```
[the remark this answers is not in the document]
Comment (mrdjen) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: think it should be ok, please check https://github.com/elastic/elasticsearch/compare/master...mrdjen:hl_client_count_api-review#diff-dd0dd6a857a14996c88dd7820d6f8c7c

**Comment by mrdjen** [comment]
Hi @hub-cap I've addressed the nits you've requested, 
Thanks

**Review by hub-cap — APPROVED** [review]

**Comment by hub-cap** [comment]
Code is merged and backported to 6.x. thanks much for the contribution! Sorry it took so long!!! :D

**Comment by mrdjen** [comment]
Hey great @hub-cap!
Np, I'm glad to contribute back!
Thank you as well and hope to contribute sometime again in the future :)

**Pull request merged**